### PR TITLE
Add `Flyout` component

### PR DESCRIPTION
**Pull request by alex-ju**
### :pushpin: Summary

Add `Flyout` component

👉 [Flyout showcase/preview (scrappy website)](https://hds-components-66d9836qe-hashicorp.vercel.app/components/flyout)
~~👉 [Flyout documentation page (new website)](https://hds-website-9sht3yqlz-hashicorp.vercel.app/components/flyout)~~ moved to #1152

### :hammer_and_wrench: Detailed description

The `Flyout` component implementation tries to stick close to the `Modal` implementation, with a few points of divergence (aside from design):
 - has a `description`, which is passed as a block to the `<F.Description>` contextual component
 - doesn't have a footer contextual component

Potential refactoring:
 - the overlay element, even a very simple atom, could be abstracted/extracted into its own utility component

### Variations
- [x] `size` variation (medium/large)
- [x] header (title, tagline, icon, description)

### Callback functions
- [x] `onOpen`
- [x] `onClose`

### :link: External links

Jira ticket: [HDS-1394](https://hashicorp.atlassian.net/browse/HDS-1394)
Figma file: https://www.figma.com/file/noyY6dUMDYjmySpHcMjhkN/branch/JOUMgI7h1ZZ9Yzvx8n3JnC/HDS-Product---Components?node-id=23645%3A53756&t=MLHoiszdlY8XlcAv-1

***

### 👀 Reviewer's checklist:

- [x] +1 Percy if applicable
- [x] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

[HDS-1394]: https://hashicorp.atlassian.net/browse/HDS-1394?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/components/app/styles/components/flyout.scss
Comment (didoo) [inline packages/components/app/styles/components/flyout.scss:96]: What do you think, would this make it easier to maintain/read/understand? (I have mixed feelings)
```suggestion
.hds-flyout {
  --max-width-responsive: calc(100vw - 40px);
  --width-calculated: min(var(--min-width-fixed), var(--max-width-responsive));
  width: var(--width-calculated);
  max-width: var(--max-width-responsive);

  &[open] {
    margin-left: calc(100% - var(--width-calculated));
  }
}

.hds-flyout--size-medium {
  --min-width-fixed: 480px;
}

.hds-flyout--size-large {
  --min-width-fixed: 720px;
}
```

Notice: variable names are tentative, and where the declarations leave in the CSS depends (may be unified in the top `.hds-flyout` block?

**Review by didoo — COMMENTED** [review]
Wow, that was easier than I expected! 😀 (I *love* the fact that you've re-used almost entirely the code from the `Modal` component! That made the code review also so much simpler 😉)

Amazing job as usual 🙌

I have added some comments inline in the code, but there are also other things I want to discuss in person (see notes below, that I've taken while reviewing). Let's set up a meeting.

## Questions/Things to discuss:

### Header (title + description)
- in modal "title" is yielded while here is `{{@title}}` - reasons why?
- in "Flyout" the `description` is yielded, while in the modal is the `title` being yielded
    - could this be confusing for the consumers?
    - need (or opportunity?) to refactor Modal?
        - 4 instances of Modal in Cloud UI https://cs.github.com/?q=Hds%3A%3AModal%20path%3A*.hbs%20repo%3Ahashicorp%2Fcloud-ui&scopeName=hashicorp&scope=org%3Ahashicorp
        - no usage in other products
- technically the "description" is not in the "header" (it's a sibling) - could it be moved to be a standalone (yielded) sub-component (eg. `SubHeader`, or `HeaderExtra` or something else), removing the need for the API differentiation between "Modal" and "Flyout"?


### Overlay

- `hasOverlay` - there is a high risk that it could be **very** controversial the decision that we don't show the overlay, but the content below is not interactive anyway (and if user clicks out of the flyout, the flyout closes). has this assumption been validated with the ambassadors (designers and engineers)?
- abstracting the overlay as "utility" component - no strong feelings on this, I'll let you decide

## Nitpicks

- `{{! template-lint-disable no-invalid-interactive }}` can be removed from `packages/components/addon/components/hds/modal/index.hbs`?

File: packages/components/tests/dummy/app/styles/pages/db-flyout.scss
Comment (alex-ju) [reply packages/components/tests/dummy/app/styles/pages/db-flyout.scss]: Indeed, we don't have plans for forms in the showcase, so not relevant – removed.

File: packages/components/tests/dummy/app/templates/components/flyout.hbs
Comment (alex-ju) [reply packages/components/tests/dummy/app/templates/components/flyout.hbs]: Switched to standard buttons

File: packages/components/tests/integration/components/hds/flyout/index-test.js
Comment (alex-ju) [reply packages/components/tests/integration/components/hds/flyout/index-test.js]: As discussed, I updated the component so we now accommodate 'rich' content for both `title` and `description` using the following API:
```hbs
<Hds::Flyout as |F|>
  <F.Header @tagline="Tagline" @icon="info">Title</F.Header>
  <F.Description>Description</F.Description>
</Hds::Flyout>
```

File: packages/components/app/styles/components/flyout.scss
Comment (alex-ju) [reply packages/components/app/styles/components/flyout.scss]: Updated to 49

File: packages/components/app/styles/components/flyout.scss
Comment (alex-ju) [reply packages/components/app/styles/components/flyout.scss]: Used a mixin here to avoid code duplication and keep the calculations in sync.

Comment (didoo) [reply packages/components/app/styles/components/flyout.scss]: Neat!

File: packages/components/tests/dummy/app/styles/pages/db-flyout.scss
Comment (alex-ju) [reply packages/components/tests/dummy/app/styles/pages/db-flyout.scss:6]: The component layout was based on a flex mode, and the showcase styling was overriding it to `inline-block`, so I switched to flex and custom height/max-height work as expected. I went with the suggested `40vh`.

File: packages/components/tests/dummy/app/templates/components/flyout.hbs
Comment (alex-ju) [reply packages/components/tests/dummy/app/templates/components/flyout.hbs]: Updated

**Review by jorytindall — APPROVED** [review]
Looks great! Awesome job, I have a follow-on PR that adds different usage scenarios, once this is merged I'll pull in the updates and fix any conflicts.

**Review by MelSumner — CHANGES_REQUESTED** [review]
There's just one issue with a header in the docs, other than that it looks great! Nice work.

**Review by didoo — APPROVED** [review]
All good for me, from the code perspective. 👍

That said, for the record, I think we should have exposed the `hasOverlay` property to allow consumers to control the presence (or not) of an overlay blocking the content below (and any interaction with it). I know @alex-ju @jorytindall have discussed this, and you came to the conclusion that it's better to not allow this (for different reasons, highlighted in the CRD). But I still think that we should have made a more pragmatic choice: I expect that sooner or later this need will come up (eg. to allow users to go through a list of items, whose details are expanded in the flyout; with the current implementation this would not be possible). Also, one of our principles is "guidance over control", and I think we should have followed it more in this case.

**Pull request merged**
